### DISCUSSION TITLE
Compiling on Platformio: How to incorporate CODE_IMAGE_STR as a build flag?

### DISCUSSION THREAD

**Discussion by jkwim** [arendst/Tasmota#20315] [Q&A] [answered]
I would like to know how to incorporate the following `CODE_IMAGE_STR` define used in `user_config_override.h` as a build flag ie.

`build_flags = ${env.build_flags} -Dxxxxxx`

    
```
    // -- CODE_IMAGE_STR is the name shown between brackets on the 
    //    Information page or in INFO MQTT messages
    #undef  :CODE_IMAGE_STR
    #define CODE_IMAGE_STR "sensorsdisplays"
```

Would 
`build_flags = ${env.build_flags} -DCODE_IMAGE_STR="sensorsdisplays"`
work?

**Comment by barbudor** [comment, marked as answer]
Not exactly
You are missing addition single quotes around the double quotes
See examples in the various ini files such as
```
[env:tasmota-zigbee]
build_flags             = ${env.build_flags}
                          -DUSE_ZIGBEE
                          -DUSE_CCLOADER
                          -DCODE_IMAGE_STR='"zigbee"'
                          -DOTA_URL='"http://ota.tasmota.com/tasmota/release/tasmota-zigbee.bin.gz"'
```

**Reply by jkwim** [reply]
Thanks for the quick answer. Will give it a go.

Greetings for the season!

**Reply by jkwim** [reply]
Well I tried but I could not get the desired results.

Looks like `user_config_override.h` has priority over `platformio_tasmota_cenv.ini`

Will try removing the default value from `user_config_override.h`.

**Reply by barbudor** [reply]
Your question was understood about moving the setting from `user_config_override.h` to `platformio_tasmota_cenv.ini`
If you leave it in `user_config_override.h`,  it "overrides" the one in `platformio_tasmota_cenv.ini`.
The file has "override" in its name for some reason...

And there is no default value in `user_config_override.h` other than the one you setup yourself

**Reply by jkwim** [reply]
I commented out the define in `user_config_override.h`

```
    // -- CODE_IMAGE_STR is the name shown between brackets on the 
    //    Information page or in INFO MQTT messages
//    #undef  :CODE_IMAGE_STR
//    #define CODE_IMAGE_STR "irraw"
```


Now the string reverts to a global value "tasmota"


`platformio_tasmota_cenv.ini`:

```
[env:tasmota-scripting-btnst-irrawsonynec]
build_flags = ${env.build_flags}
         -DFIRMWARE_SCRIPTING_BUTTONSTATE_IRRAW
         -DCODE_IMAGE_STR='"scripting_btnst__irraw_sonynec"'
         -D_IR_ENABLE_DEFAULT_=false
         -DSEND_RAW=true
         -DSEND_GLOBALCACHE=true
         -DDECODE_GLOBALCACHE=true
         -DSEND_HASH=true
         -DDECODE_HASH=true
         -DDECODE_SONY=true
         -DSEND_SONY=true
         -DDECODE_NEC=true
         -DSEND_NEC=true
```

Wondering where the mistake is.

**Reply by barbudor** [reply]
Not knowing what is hidden behind `-DFIRMWARE_SCRIPTING_BUTTONSTATE_IRRAW` makes it difficult to help

But with an totally empty `user_config_override.h` and this in my `platformio_tasmota_cenv.ini`:
```
[env:tasmota-test]
build_flags             = ${env.build_flags} -DCODE_IMAGE_STR='"barbudor"'
```

I get the expected result:
```
00:00:00.077 Project tasmota - nodemcu Version 13.3.0.1(barbudor)-2_7_4_9(2023-12-26T16:36:46)
```
![image](https://github.com/arendst/Tasmota/assets/2996042/245a1ffa-a20d-42af-9a80-35025960e5de)

**Comment by jkwim** [comment]
It contains my other #define statements.

I am trying to get limited number of IR drivers in to scripting compilation. Specifically I want a mechanism to build multiple driver combinations easily.

The builds themselves work fine.

It is just that I need to be able to clearly distinguish which one is what by looking at the  CODE_IMAGE_STR.

May be the string that I use is too long?

Will try shorter strings and see whether it works.
 

```
#ifdef FIRMWARE_SCRIPTING_BUTTONSTATE_IRRAW
    // This line will issue a warning during the build (yellow in
    // VSCode) so you see which section is used
    #warning **** Build: SCRIPTING BUTTONSTATE IRRAW ****
.
.
.


        #ifndef USE_IR_REMOTE_FULL
          #define USE_IR_REMOTE_FULL  // Support all IR protocols from IRremoteESP8266
        #endif

 
#endif  // FIRMWARE_SCRIPTING_BUTTONSTATE_IRRAW
```

**Reply by jkwim** [reply]
String length does not seemed to be an issue.

I can set the full string that I wanted via `user_config_override.h` without any issue.

![image](https://github.com/arendst/Tasmota/assets/84637540/24825588-63b4-481f-9a2f-fccc09eb2c15)

Setting it via build flag does not work no matter.

I even tried leaving `#undef  CODE_IMAGE_STR` in `user_config_override.h` then the compilation fails citing undefined `CODE_IMAGE_STR`. That suggests that in my setup the flag does not get recognized at all.

The only thing left is to see whether the order of flags matter. I did try moving this flag to the top and the bottom but no progress.

Running out of ideas :-)

**Comment by jkwim** [comment]
Tried a build like this:

```
[env:tasmota-irraw-test]
build_flags = ${env.build_flags}
;         -DFIRMWARE_SCRIPTING_BUTTONSTATE_IRRAW
         -DCODE_IMAGE_STR='"irraw-test"'
;         -D_IR_ENABLE_DEFAULT_=false
;         -DSEND_RAW=true
;         -DSEND_GLOBALCACHE=true
;        -DDECODE_GLOBALCACHE=true
;         -DSEND_HASH=true
;         -DDECODE_HASH=true
;         -DDECODE_SONY=true
;         -DSEND_SONY=true
;         -DDECODE_NEC=true
;         -DSEND_NEC=true

lib_extra_dirs              = lib/lib_basic, lib/lib_ssl, lib/lib_div
```

However the result was default string "tasmota"

**Comment by jkwim** [comment]
Just for curiosity I copied your configure directly in to `platformio_tasmota_cenv.ini `but it did not work. So somewhere in my setup the build option is not getting picked up.


```
[env:tasmota-test]
build_flags             = ${env.build_flags} -DCODE_IMAGE_STR='"barbudor"'
```

**Comment by Jason2866** [comment]
You overlook something doing harm in your setup. Without providing **all** Platformio files you have changes and the `user_config_override.h` we can't help.
